README

*** NOTE ***
============

This has commits from contributors to master before I took a deep breath and renamed branches.  The pyspf-2.1 branch is 
an earlier snapshot.  The goal of this branch is to merge fixes from master (now pyspf-2.0) with new ideas for pyspf-2.1.

 1. Use a directory for the module.  
 2. use dnsplug, dnsswitch, or other separate module, perhaps under spf directory for selecting dns library.  dkimpy also 
 needs the pluggable DNS support.
 3. modules for ip6 support on ancient python system - probably not terribly relevant anymore.

SPF
Sender-Policy-Framework queries in Python.

Quick Start
===========

Installation
------------
This package requires PyDNS (or Py3DNS for running with Python 3) and either
the ipaddr or python3.3 and later.  PyDNS is available at
http://pydns.sourceforge.net.  Binary and source RPMs for PyDNS are also
available from http://pymilter.sourceforge.net.  Py3DNS is available on pypi
and at https://launchpad.net/py3dns.  The ipaddr module is available from
http://code.google.com/p/ipaddr-py or as part of the Python standard library
starting with python3.3 (as ipaddress).  This package requires authres from
either pypi or http://launchpad.net/authentication-results-python to process
and generate RFC 5451 Authentication Results headers.

After unpacking the source distribution, install this in your site-
specific Python extension directory::

	% python setup.py build
	% su
	# python setup.py install

The minimum Python version required is python2.6.  The spf module in this
version has been tested with python3.2 and does not require using 2to3. It
will work with all versions of pydns or py3dns.  It works either with the
stand alone ipaddr module or the standard library ipaddress module.

Testing
-------
After this package is installed, cd into the test directory and
execute testspf.py::

    % cd test
    % python testspf.py
    WARN: invalid-domain-long in rfc4408-tests.yml, 8.1/2, 5/10: fail preferred to temperror
    WARN: txttimeout in rfc4408-tests.yml, 4.4/1: fail preferred to temperror
    WARN: spfoverride in rfc4408-tests.yml, 4.5/5: pass preferred to fail
    WARN: multitxt1 in rfc4408-tests.yml, 4.5/5: pass preferred to permerror
    WARN: multispf2 in rfc4408-tests.yml, 4.5/6: permerror preferred to pass
    ..
    ----------------------------------------------------------------------
    Ran 2 tests in 3.036s

    OK

This runs the SPF council test-suite as of when this package was built.
It does not test the pyDNS installation, but uses an internal driver.
This avoids changing results due to DNS timeouts.

In addition, spf.py runs an internal self-test every time it is used from the
command line.

If you're running on Mac OS X, and it looks like DNS.DiscoverNameServers()
is failing, you'll need to edit your /etc/resolv.conf and specify a
domain name.  For some reason, OS X writes out resolv.conf with a single
'domain' line, which isn't good at all.  Later versions of py3dns have been
updated to better support Max OS X.


Description
===========
SPF does email sender validation.  For more information about SPF,
please see http://www.openspf.org

One incompatible change was introduced in version 1.7.  Prior to version 1.7,
connections from a local IP address (127...) would always return a Pass 
result.  The special case was eliminated.  Programs calling pySPF should not
do SPF checks on locally submitted mail.

This SPF client is intended to be installed on the border MTA, checking
if incoming SMTP clients are permitted to forward mail.  The SPF check
should be done during the MAIL FROM:<...> command.

There are two ways to use this package.  The first is from the command
line::
	% python spf.py {ip-addr} {mail-from} {helo}

For instance, during an SMTP exchange from client 69.55.226.139::
	S: 220 mail.example.com ESMTP Postfix
	C: EHLO mx1.wayforward.net
	S: 250-mail.example.com
	S: ...
	S: 250 8BITMIME
	C: MAIL FROM:<terry@wayforward.net>

Then the following command line would check if this is a valid sender:
	% ./spf.py 69.55.226.139 terry@wayforward.net mx1.wayforward.net
	('pass', 250, 'sender SPF authorized')

Command line calls return RFC 4408/7208 result codes, i.e. 'pass', 'fail',
'neutral', 'softfail, 'permerror', or 'temperror'.

The second way is via the module's APIs.

The legacy (pySPF 1.6) API:
	>>> import spf
	>>> spf.check(i='69.55.226.139',
	...           s='terry@wayforward.net',
	...           h='mx1.wayforward.net')
	('pass', 250, 'sender SPF authorized')

The first element in the tuple is one of 'pass', 'fail', 'netural', 'softfail',
'unknown', or 'error'.  The second is the SMTP response status code: 550 for 
'fail', 450 for 'error' and 250 for all else.  The third is an explanation.

Note: SPF results alone are never sufficient to decide that a message should be
accepted.  Accept, reject, or defer decisions are a function of local reciever
policy.

The RFC 4408/7208 compliant API:
        >>> import spf
        >>> spf.check2(i='69.55.226.139',
        ...           s='terry@wayforward.net',
        ...           h='mx1.wayforward.net')
        ('pass', 'sender SPF verified')

The first element in the tuple is one of 'pass', 'fail', 'neutral', 'softfail,
'permerror', or 'temperror'.  The second is an explanation.

This package also provides two additional helper scripts; type99.py and 
spfquery.py.  The type99.py script will convert DNS TXT strings to a binary 
equivalent suitable for use in a BIND zone file.  The spfquery.py script is a
Python reimplementination of Wayne Schlitt's spfquery command line tool.

The type99.py script is called from the command line as follows:

python type99.py "v=spf1 -all" {Note: Use your desired SPF record instead.}
\# 12 0b763d73706631202d616c6c {This is the correct result for "v=spf1 -all"}

or 

python type99 - {File name}

The input file format is a standard BIND Zone file.  The type99 script will add
a Type99 record for each TXT record found in the file.  Use of DNS type 99
(type SPF) was removed from SPF in RFC 7208, so this script should be of
historical interest only.

The spfquery.py script is called with a number of possible options.  Options can
either use standard '-' prefix or be PERL style long options, '--'.  Supported
options are:

"--file" or "-file" {filename}: Read the query (or queries) from the designated 
    file.  If {filename} is '0', then query inputs are read from STDIN.

 "--ip" or "-ip" {address}: Client IP address to use for SPF check.


"--sender" or "-sender" {Mail From address}: Envelope sender from which mail was
    received.

"--helo" or "-helo" {client hostname}: HELO/EHLO name used by SMTP client.

"--local" or "-local" {local policy SPF string}: Additional SPF mechanisms to be
    checked on the basis of local policy.  Note that local policy matches are 
    not strictly SPF results.  Local policy processing is not defined in RFC 
    4408 or RFC 7208.  Result may vary among SPF implementations.

"--rcpt-to" or "rcpt-to" {rcpt-to address - if available}: Receipt to address is
    not used for actual SPF processing, but if available it can be useful for 
    logging, spf-received header construction, and providing useful rejection
    messages when messages are rejected due to SPF.

--default-explanation" or "-default-explanation" {explanation string}: Default
    Fail explanation string to be used.

"--sanitize" or "-sanitize" and "--debug" or "-debug": These options are no-op
    in the Python implementation, but are valid inputs to provide compatibliity
    with input files developed to work with the original PERL and C spfquery
    implementations.

Overall per SPF check time limits can be controlled by passing querytime
to the spf.check2 function or when initializing a spf.query object.
It is set to 20 seconds by default based on RFC 7208.  If querytime is set to
0, then the overall time limit is disabled and the per DNS lookup limit is used
instead.  This defaults to 20 seconds and can be controlled via
spf.MAX_PER_LOOKUP_TIME.  RFC 4408 says that the overall limit MAY be used and
recommends no less than 20 seconds if it is. RFC 7208 is stronger, so a
default limit aligned to the RFC requirements is now used.

The type99.py script is called from the command line as follows:

python type99.py "v=spf1 -all" {Note: Use your desired SPF record instead.}
\# 12 0b763d73706631202d616c6c {This is the correct result for "v=spf1 -all"}

or 

python type99 - {File name}

The input file format is a standard BIND Zone file.  The type99 script will add
a Type99 record for each TXT record found in the file.

The spfquery.py script is called with a number of possible options.  Options can
either use standard '-' prefix or be PERL style long options, '--'.  Supported
options are:

"--file" or "-file" {filename}: Read the query (or queries) from the designated 
    file.  If {filename} is '0', then query inputs are read from STDIN.

 "--ip" or "-ip" {address}: Client IP address to use for SPF check.


"--sender" or "-sender" {Mail From address}: Envelope sender from which mail was
    received.

"--helo" or "-helo" {client hostname}: HELO/EHLO name used by SMTP client.

"--local" or "-local" {local policy SPF string}: Additional SPF mechanisms to be
    checked on the basis of local policy.  Note that local policy matches are 
    not strictly SPF results.  Local policy processing is not defined in RFC 
    4408.  Result may vary among SPF implementations.

"--rcpt-to" or "rcpt-to" {rcpt-to address - if available}: Receipt to address is
    not used for actual SPF processing, but if available it can be useful for 
    logging, spf-received header construction, and providing useful rejection
    messages when messages are rejected due to SPF.

--default-explanation" or "-default-explanation" {explanation string}: Default
    Fail explanation string to be used.

"--sanitize" or "-sanitize" and "--debug" or "-debug": These options are no-op
    in the Python implementation, but are valid inputs to provide compatibliity
    with input files developed to work with the original PERL and C spfquery
    implementations.

License: Python Software Foundation License

Author:
Terence Way terry@wayforward.net
http://www.wayforward.net/spf/

Maintainers:
Stuart Gathman stuart@gathman.org
Scott Kitterman scott@kitterman.com
http://cheeseshop.python.org/pypi/pyspf

Code is currently hosted at https://github.com/sdgathman/pyspf/


CHANGELOG
Version 2.1.0
  * Switch to setuptools because "it's the future"
  * Require ipaddress instead of ipaddr for python2
  * Fix issue Getting list of ip networks fails if prefix is 32 or more,
    GitHub #24

Version 2.0.14 - January 1, 2020
  * Fix doctest for CNAME fixes to work with python and python3
  * Fix dnspython integration so that SPF TempError is properly raised when
    there are timeout or no nameserver errors
  * Add missing use of timeout parameter for dnspython DNSLookup
  * Restore DNSLookup API for pydnsv(DNS) for tcp fallback works again
  * Update Installation section of README.md

Version 2.0.13 - September 2, 2019
  * Add support for use of dnspython (dns) if installed
  * Catch ValueError due to improper IP address in connect IP or in ip4/ip6
    mechanisms
  * Fix for CNAME processing causing incorrect permerrors

Version 2.0.12 - August 5, 2015
  * Reset void_lookups at top of check()
  * Ignore permerror for best_guess()
  * Don't crash on null DNS TXT record (ignore): test case null-text
  * Trailing spaces are allowed by 4.5/2: test case trailing-space
  * Make CNAME loop result in unknown host: test case ptr-cname-loop
  * Test case and fix for mixed case CNAME loop, test case ptr-cname-loop

Version 2.0.11 - December 5, 2014
  * Fix another bug in SPF record parsing that caused records with terms
    separated by multple spaces as invalid, but they are fine per the ABNF
  * Downcase names in additional answers returned by DNS before adding
    to cache, since case inconsistency can cause PTR match failures (initial
    patch thanks to Joni Fieggen) and other problems.

Version 2.0.10 - September 2, 2014
  * Fix bug in SPF record parsing that caused all 'whitespace' characters to
    be considered valid term separators and not just spaces
  * Fixed multiple bugs in temperror processing that would lead to tracebacks
    instead of correct error processing
  * Fix AAAA not flagged as bytes when strict=2 
  * Include '~' as safe char in url quoted macro expansion

Version 2.0.9 - April 29, 2014
  * Update for new SPF standards track RFC 7208
    - Add processing for new void lookups processing limit
    - Default SPF process timeout limit to 20 seconds per RFC 7208 4.6.4
    - Change default DNS timeout to 20 seconds in DNSLookup to better match
      RFC 7208 4.6.4
    - Make mx lookups > 10 a permerror per RFC 7208 and mx-limit test
    - Add RFC 7208 specific test suite and make allowance for RFC 7208 changes
      in RFC 4408 test suite
    - Convert YAML tests to TestCases, and have testspf.py return success/fail.

Version 2.0.8 - July 24, 2013
  * Use ipaddr/ipaddres module in place of custom IP processing code
  * Numerous python3 compatibility fixes
  * Improved unicode error detection in SPF records
  * Fixed a bug caused by a null CNAME in cache

Version 2.0.7 - January 19, 2012
  * Allow for timeouts to be global for all DNS lookups instead of per DNS lookup
    to allow for MAY processing time limitsin RFC 4408 10.1.  See README for
    details.
  * Use openspf.net for SPF web site instead of openspf.org
  * Extend query.get_header to return either Received-SPF (still default) or
    RFC 5451 Authentication Results headers (needs authres 0.3 or later)
  * Rework query.parse_header:
    - Make query.parse_header automatically select Received-DPF or
      Authentication Results header types and use them to collect SPF
      results from trusted relays
    - Add query.parse_header_spf and query.parse_header_ar functions for
      header type specific processing
  * Finish Python3 port - works with python2.6/2.7/3.2 and 2to3 is no longer
    required - will also work with newer py3dns where TXT records are returned
    as type bytes and not strings
  * Accounts for new py3dns error classes coming in py3dns 3.0.2 (but fully
    backward compatible with earlier versions)
  * check for 7-bit ascii on TXT and SPF records
  * fix CNAME chain duplicating TXT records

Version 2.0.6 - October 27, 2011
  * Refactor code so that 2to3 will provide a working python3 module - Now
    requires at least python2.6
  * Update spfquery.py, type99.py, and testspf.py to work with either python or
    python3 (2to3 not needed for these scripts)
    - SPF test suite can now be run from either python or python3
  * Ensure Temperror for all DNS rcodes other than 0 and 3 per RFC 4408
  * Parse Received-SPF header
  * Report CIDR error only for valid mechanism
  * Handle invalid SPF record on command line
  * Add timeout to check2

Version 2.0.5 - July 29, 2008
  * Add TCP fallback if DNS UDP reply is truncated
    - Fixes inconsistent results from trying to use partial UDP replies
  * Correct Received-SPF formatting
  * Minor updates to reflect RFC 4408 errata
  * Added License file for RFC 4408 test suite
  * Update RFC 4408 test suite from svn
  * Fix Type99 conversion script to work with multi-string TXT records
  * Timeout parameter

Version 2.0.4 - January 24, 2007
  * Correct unofficial 'best guess' processing.
  * PTR validation processing cleanup
  * Improved detection of exp= errors
  * Keyword parameters on get_header()

Version 2.0.3 - January 15, 2007
  * IPv6 compatibility test fix to support Python 2.2
  * Change DNS queries to only check Type SPF in Harsh mode
  * pyspf requires pydns, python-pyspf requires python-pydns
  * Record matching mechanism and add to Received-SPF header.
  * Test for RFC4408 6.2/4, and fix spf.py to comply.
  * Permerror for more than one exp or redirect modifier.
  * Parse op= modifier

Version 2.0.2 - January 4, 2007
  * Update openspf URLs
  * Update Readme to better describe available pyspf interfaces
  * Add basic description of type99.py and spfquery.py scripts
  * Add usage instructions for type99.py DNS RR type conversion script
  * Add spfquery.py usage instructions
  * Incorporate downstream feedback from Debian packager
  * Fix key-value quoting in get_header

Version 2.0.1 - December 08, 2006 

pyspf/spfquery.py
#!/usr/bin/python

# Author: Stuart D. Gathman <stuart@bmsi.com>
# Copyright 2004 Business Management Systems, Inc.

# This module is free software, and you may redistribute it and/or modify
# it under the same terms as Python itself, so long as this copyright message
# and disclaimer are retained in their original form.

# Emulate the spfquery command line tool used by Wayne Schlitt's SPF test suite

# $Log$
# Revision 1.4.2.3  2011/10/27 04:44:58  kitterma
# Update spfquery.py to work with 2.6, 2.7, and 3.2:
#  - raise ... as ...
#  - print()
#
# Revision 1.4.2.2  2008/03/26 14:34:35  kitterma
# Change shebangs to #!/usr/bin/python throughout.
#
# Revision 1.4.2.1  2006/12/23 05:31:22  kitterma
# Minor updates for packaging lessons learned from Ubuntu
#
# Revision 1.4  2006/11/20 18:39:41  customdesigned
# Change license on spfquery.py.  Update README.  Move tests to test directory.
#
# Revision 1.3  2005/07/22 02:11:57  customdesigned
# Use dictionary to check for CNAME loops.  Check limit independently for
# each top level name, just like for PTR.
#
# Revision 1.2  2005/07/14 04:18:01  customdesigned
# Bring explanations and Received-SPF header into line with
# the unknown=PermErr and error=TempErr convention.
# Hope my case-sensitive mech fix doesn't clash with Scotts.
#
# Revision 1.1.1.1  2005/06/20 19:57:32  customdesigned
# Move Python SPF to its own module.
#
# Revision 1.2  2005/06/02 04:18:55  customdesigned
# Update copyright notices after reading article on /.
#
# Revision 1.1.1.1  2005/05/31 18:07:19  customdesigned
# Release 0.6.9
#
# Revision 2.3  2004/04/19 22:12:11  stuart
# Release 0.6.9
#
# Revision 2.2  2004/04/18 03:29:35  stuart
# Pass most tests except -local and -rcpt-to
#
# Revision 2.1  2004/04/08 18:41:15  stuart
# Reject numeric hello names
#
# Driver for SPF test system

import spf
import sys

from optparse import OptionParser

class PerlOptionParser(OptionParser):
    def _process_args (self, largs, rargs, values):
        """_process_args(largs : [string],
                         rargs : [string],
                         values : Values)

        Process command-line arguments and populate 'values', consuming
        options and arguments from 'rargs'.  If 'allow_interspersed_args' is
        false, stop at the first non-option argument.  If true, accumulate any
        interspersed non-option arguments in 'largs'.
        """
        while rargs:
            arg = rargs[0]
            # We handle bare "--" explicitly, and bare "-" is handled by the
            # standard arg handler since the short arg case ensures that the
            # len of the opt string is greater than 1.
            if arg == "--":
                del rargs[0]
                return
            elif arg[0:2] == "--":
                # process a single long option (possibly with value(s))
                self._process_long_opt(rargs, values)
            elif arg[:1] == "-" and len(arg) > 1:
                # process a single perl style long option
                rargs[0] = '-' + arg
                self._process_long_opt(rargs, values)
            elif self.allow_interspersed_args:
                largs.append(arg)
                del rargs[0]
            else:
                return

def format(q):
  res,code,txt = q.check()
  print(res)
  if res in ('pass','neutral','unknown'): print()
  else: print(txt)
  print('spfquery:',q.get_header_comment(res))
  print('Received-SPF:',q.get_header(res,'spfquery'))

def main(argv):
  parser = PerlOptionParser()
  parser.add_option("--file",dest="file")
  parser.add_option("--ip",dest="ip")
  parser.add_option("--sender",dest="sender")
  parser.add_option("--helo",dest="hello_name")
  parser.add_option("--local",dest="local_policy")
  parser.add_option("--rcpt-to",dest="rcpt")
  parser.add_option("--default-explanation",dest="explanation")
  parser.add_option("--sanitize",type="int",dest="sanitize")
  parser.add_option("--debug",type="int",dest="debug")
  opts,args = parser.parse_args(argv)
  if opts.ip:
    q = spf.query(opts.ip,opts.sender,opts.hello_name,local=opts.local_policy)
    if opts.explanation:
      q.set_default_explanation(opts.explanation)
    format(q)
  if opts.file:
    if opts.file == '0':
      fp = sys.stdin
    else:
      fp = open(opts.file,'r')
    for ln in fp:
      ip,sender,helo,rcpt = ln.split(None,3)
      q = spf.query(ip,sender,helo,local=opts.local_policy)
      if opts.explanation:
        q.set_default_explanation(opts.explanation)
      format(q)
    fp.close()
    
if __name__ == "__main__":
  import sys
  main(sys.argv[1:])
